internal/pkg/keybindings/globalhandlers.go

```diff
@@ -200,44 +214,112 @@ func (h *ClearPendingDeleteHandler) Fn() func(g *gocui.Gui, v *gocui.View) error
 	return func(g *gocui.Gui, v *gocui.View) error {
-		h.commandPanelWidget.ToggleShowHide()
+		keyBindings := GetKeyBindingsAsStrings()
+
+		paletteWidth := h.commandPanelWidget.Width() - 4
+		options := []views.CommandPanelListOption{}
+		for _, command := range h.commands {
+			if command.IsEnabled() {
+				commandID := command.ID()
+				commandDisplayText := command.DisplayText()
+
+				// ensure binding is in upper-case
+				binding := keyBindings[commandID]
+				for i, b := range binding {
+					binding[i] = strings.ToUpper(b)
+				}
+
+				// calculate padding to right-align shortcut
+				bindingString := fmt.Sprintf("%s", binding)
+				padAmount := paletteWidth - len(commandDisplayText) - len(bindingString)
+				if padAmount < 0 {
+					padAmount = 0 // TODO - we should also look at truncating the DisplayText
+				}
+
+				option := views.CommandPanelListOption{
+					ID:          commandID,
+					DisplayText: command.DisplayText() + strings.Repeat(" ", padAmount) + bindingString,
+				}
+
+				options = append(options, option)
+			}
+		}
+		h.commandPanelWidget.ShowWithText("Command Palette", "", &options, h.CommandPanelNotification)
 		return nil
 	}
 }
 
+func (h *OpenCommandPanelHandler) CommandPanelNotification(state views.CommandPanelNotification) {
+	if state.EnterPressed {
+		h.commandPanelWidget.Hide()
+		for _, command := range h.commands {
+			if command.ID() == state.SelectedID {
+				// invoke via Update to allow Hide to restore preview view state
+				h.gui.Update(func(gui *gocui.Gui) error {
+					return command.Invoke()
+				})
+				return
+			}
+		}
+	}
+}
+
 ////////////////////////////////////////////////////////////////////
 
 ////////////////////////////////////////////////////////////////////
 type CommandPanelFilterHandler struct {
 	GlobalHandler
 	commandPanelWidget *views.CommandPanelWidget
+	list               *views.ListWidget
 }
 
-func NewCommandPanelFilterHandler(commandPanelWidget *views.CommandPanelWidget) *CommandPanelFilterHandler {
+var _ Command = &CommandPanelFilterHandler{}
+
+func NewCommandPanelFilterHandler(commandPanelWidget *views.CommandPanelWidget, list *views.ListWidget) *CommandPanelFilterHandler {
 	handler := &CommandPanelFilterHandler{
 		commandPanelWidget: commandPanelWidget,
+		list:               list,
 	}
 	handler.id = HandlerIDFilter
 	return handler
 }
 
 func (h *CommandPanelFilterHandler) Fn() func(g *gocui.Gui, v *gocui.View) error {
 	return func(g *gocui.Gui, v *gocui.View) error {
-		h.commandPanelWidget.ShowWithText("/")
-		return nil
+		return h.Invoke()
+	}
+}
+func (h *CommandPanelFilterHandler) DisplayText() string {
+	return "Filter"
+}
+func (h *CommandPanelFilterHandler) IsEnabled() bool {
+	return true
+}
+func (h *CommandPanelFilterHandler) Invoke() error {
+	h.commandPanelWidget.ShowWithText("Filter", "", nil, h.CommandPanelNotification)
+	return nil
+}
+func (h *CommandPanelFilterHandler) CommandPanelNotification(state views.CommandPanelNotification) {
```

Comment:
:) My brain is mapping this to somewhere between `INotifyPropertyChanged` and `ICommand` :+1: